File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        
        l, r  = 0, len(height) - 1
        leftMax = height[l]
        rightMax = height[r]
        res = 0
        
        while l < r:
            if leftMax < rightMax:
                l += 1
                leftMax = max(leftMax, height[l])
                res += leftMax - height[l]
            else:
                r -= 1
                rightMax = max(rightMax, height[r])
                res += rightMax - height[r]
                
        return res

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int maxleft = -1, maxright = -1;
        int water = 0;
        int l = 0, r = height.length - 1;
        
        while (l <= r){
            maxleft = Math.max(maxleft, height[l]);
            water += maxleft - height[l];
            
            maxright = Math.max(maxright, height[r]);
            water += maxright - height[r];
            
            if (maxleft < maxright) l++;
            else r--;
        }
        return water;
        
    }
}